types.go

```diff
@@ -31,3 +41,259 @@ func PHPString(s string, persistent bool) unsafe.Pointer {
 
 	return unsafe.Pointer(zendStr)
 }
+
+// PHPKeyType represents the type of PHP hashmap key
+type PHPKeyType int
+
+const (
+	PHPIntKey PHPKeyType = iota
+	PHPStringKey
+)
+
+type PHPKey struct {
+	Type PHPKeyType
+	Str  string
+	Int  int64
+}
+
+// Array represents a PHP array with ordered key-value pairs
+type Array struct {
+	keys   []PHPKey
+	values []interface{}
+}
+
+// SetInt sets a value with an integer key
+func (arr *Array) SetInt(key int64, value interface{}) {
+	arr.keys = append(arr.keys, PHPKey{Type: PHPIntKey, Int: key})
+	arr.values = append(arr.values, value)
+}
+
+// SetString sets a value with a string key
+func (arr *Array) SetString(key string, value interface{}) {
+	arr.keys = append(arr.keys, PHPKey{Type: PHPStringKey, Str: key})
+	arr.values = append(arr.values, value)
+}
+
+// Append adds a value to the end of the array with the next available integer key
+func (arr *Array) Append(value interface{}) {
+	nextKey := arr.getNextIntKey()
+	arr.SetInt(nextKey, value)
+}
+
+// getNextIntKey finds the next available integer key
+func (arr *Array) getNextIntKey() int64 {
+	maxKey := int64(-1)
+	for _, key := range arr.keys {
+		if key.Type == PHPIntKey && key.Int > maxKey {
+			maxKey = key.Int
+		}
+	}
+
+	return maxKey + 1
+}
+
+// Len returns the number of elements in the array
+func (arr *Array) Len() uint32 {
+	return uint32(len(arr.keys))
+}
+
+// At returns the key and value at the given index
+func (arr *Array) At(index uint32) (PHPKey, interface{}) {
+	if index >= uint32(len(arr.keys)) {
+		return PHPKey{}, nil
+	}
+	return arr.keys[index], arr.values[index]
+}
+
+// EXPERIMENTAL: GoArray converts a zend_array to a Go Array
+func GoArray(arr unsafe.Pointer) *Array {
+	result := &Array{
+		keys:   make([]PHPKey, 0),
+		values: make([]interface{}, 0),
+	}
+
+	if arr == nil {
+		return result
+	}
+
+	zval := (*C.zval)(arr)
+	hashTable := (*C.HashTable)(castZval(zval, C.IS_ARRAY))
+
+	if hashTable == nil {
+		return result
+	}
+
+	used := hashTable.nNumUsed
+	if htIsPacked(hashTable) {
+		for i := C.uint32_t(0); i < used; i++ {
+			v := C.get_ht_packed_data(hashTable, i)
+			if v != nil && C.zval_get_type(v) != C.IS_UNDEF {
+				value := convertZvalToGo(v)
+				result.SetInt(int64(i), value)
+			}
+		}
+	} else {
+		for i := C.uint32_t(0); i < used; i++ {
+			bucket := C.get_ht_bucket_data(hashTable, i)
+			if bucket != nil && C.zval_get_type(&bucket.val) != C.IS_UNDEF {
+				v := convertZvalToGo(&bucket.val)
+
+				if bucket.key != nil {
+					keyStr := GoString(unsafe.Pointer(bucket.key))
+					result.SetString(keyStr, v)
+				} else {
+					result.SetInt(int64(bucket.h), v)
+				}
+			}
+		}
+	}
+
+	return result
+}
+
+// PHPArray converts a Go Array to a PHP zend_array.
+func PHPArray(arr *Array) unsafe.Pointer {
+	if arr == nil || arr.Len() == 0 {
+		return unsafe.Pointer(createNewArray(0))
+	}
+
+	isList := true
+	for i, k := range arr.keys {
+		if k.Type != PHPIntKey || k.Int != int64(i) {
```

Comment:
will this be an issue on 32bit machines where the keys will be 32 bits?

Reply:
It should be good, manipulating int64 on 32 bits system is just a bit slower. But we can also type it with `int` instead of `int64`, so it adapts depending on the system. Not sure which would be better, given that 32 bits system are becoming pretty rare

Reply:
This actually came up on php-internals recently. It is less rare than you'd think. For example, wasm implementations are 32 bit.